### PR TITLE
cmake: Improve code signing behavior on macOS

### PR DESCRIPTION
**Pull request by PatTheMav**
### Description
~Fixed issues on Linux:~

* ~libobs and obs-frontend-api headers are not installed by default~
* ~RPATH for frontend-tools was malformed~

Improved behaviour on macOS:
* When a valid team ID is provided and the Xcode generator is used, automatic code-signing is enabled for all targets in the project
* Different deployment targets are set in Xcode for arm64 and Intel x86 when universal builds are configured

Other improvements:
* Support for legacy obs-browser submodule removed, because it is not necessary anymore
* Update plugin helper CMake script shipped in repository with more recent variant adapted from obs-plugintemplate

### Motivation and Context
~The Linux fixes enable easier creation of distributable packages for maintainers and ensure that frontend-tools can find the scripting plugin if available.~
*Linux fixes have been moved into their own PR.*

The macOS changes allow maintainers to use a team ID with an active Apple Developer membership to use proper code signing during development.

### How Has This Been Tested?
OBS built and ran on macOS 12.5, Ubuntu 22, and Windows 11.

### Types of changes
- Bug fix (non-breaking change which fixes an issue)
- Tweak (non-breaking change to improve existing functionality)
- Code cleanup (non-breaking change which makes code smaller or more readable)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: cmake/external/ObsPluginHelpers.cmake [hunk 1]
```diff
@@ -51,65 +54,344 @@ if(NOT CMAKE_BUILD_TYPE)
...
-    # install_name_tool preserves ad-hoc signatures. This option is supported by
-    # `codesign` on macOS 11 or higher.
-    #
-    # See CMake Issue 21854:
-    # (https://gitlab.kitware.com/cmake/cmake/-/issues/21854).
-
-    set(CMAKE_XCODE_GENERATE_SCHEME ON)
-    if(OBS_CODESIGN_LINKER)
-      set(CMAKE_XCODE_ATTRIBUTE_OTHER_CODE_SIGN_FLAGS "-o linker-signed")
+  # Set up codesigning for Xcode builds with team ids or standalone builds with
```
Comment (RytoEX) [inline cmake/external/ObsPluginHelpers.cmake:330]: ids -> IDs

File: cmake/external/ObsPluginHelpers.cmake [hunk 2]
```diff
@@ -51,65 +54,344 @@ if(NOT CMAKE_BUILD_TYPE)
...
+  if(NOT OBS_BUNDLE_CODESIGN_TEAM)
+    if(NOT OBS_BUNDLE_CODESIGN_IDENTITY)
+      set(OBS_BUNDLE_CODESIGN_IDENTITY
+          "-"
+          CACHE STRING "OBS code signing identity for macOS" FORCE)
+    endif()
+    set(CMAKE_XCODE_ATTRIBUTE_CODE_SIGN_IDENTITY
+        ${OBS_BUNDLE_CODESIGN_IDENTITY})
+  else()
+    # Team id specified, warn if Xcode generator is not used and fall back to
```
Comment (RytoEX) [inline cmake/external/ObsPluginHelpers.cmake:341]: id -> ID

File: cmake/external/ObsPluginHelpers.cmake [hunk 3]
```diff
@@ -51,65 +54,344 @@ if(NOT CMAKE_BUILD_TYPE)
...
+      unset(OBS_BUNDLE_CODESIGN_IDENTITY)
+      set_property(CACHE OBS_BUNDLE_CODESIGN_TEAM
+                   PROPERTY HELPSTRING "OBS code signing team for macOS")
+      set(CMAKE_XCODE_ATTRIBUTE_CODE_SIGN_STYLE Automatic)
+      set(CMAKE_XCODE_ATTRIBUTE_DEVELOPMENT_TEAM ${OBS_BUNDLE_CODESIGN_TEAM})
     endif()
   endif()
 
+  # Set path to entitlements property list for codesigning (entitlements should
+  # match the host binary, in this case OBS.app)
```
Comment (RytoEX) [inline cmake/external/ObsPluginHelpers.cmake:363]: Drop the parenthesis. Use two sentences.

File: cmake/external/ObsPluginHelpers.cmake [hunk 4]
```diff
@@ -51,65 +54,344 @@ if(NOT CMAKE_BUILD_TYPE)
...
+  set(OBS_CODESIGN_ENTITLEMENTS
+      ${CMAKE_SOURCE_DIR}/cmake/bundle/macos/entitlements.plist
+      CACHE INTERNAL "Path to codesign entitlements plist")
+  # Enable linker codesigning by default - building OBS or plugins on host
+  # systems older than macOS 10.15 is not supported
```
Comment (RytoEX) [inline cmake/external/ObsPluginHelpers.cmake:368]: Drop the hyphen. Use two sentences.

File: cmake/external/ObsPluginHelpers.cmake [hunk 5]
```diff
@@ -263,11 +564,71 @@ else()
...
+      #   marked for no inlining, see
+      #   https://docs.microsoft.com/en-us/cpp/build/reference/ob-inline-function-expansion?view=msvc-170
+      # * Enable UNICODE support, see
+      #   https://docs.microsoft.com/en-us/windows/win32/learnwin32/working-with-strings#unicode-and-ansi-functions
+      # * DISABLE warnings about using POSIX function names,
+      #   https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-3-c4996?view=msvc-170#posix-function-names
+      # * DISABLE warnings about unsafe CRT library functions, see
+      #   https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-3-c4996?view=msvc-170#unsafe-crt-library-functions
+      # * DISABLE warnings about nonstandard nameless structs/unions, see
+      #   https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-4-c4201?view=msvc-170
```
Comment (RytoEX) [inline cmake/external/ObsPluginHelpers.cmake:587]: Could probably drop the word "see" and just end the comment, unless there is some danger of cmake-format wrapping poorly.

File: cmake/external/ObsPluginHelpers.cmake [hunk 6]
```diff
@@ -263,11 +564,71 @@ else()
...
+      #   https://docs.microsoft.com/en-us/cpp/build/reference/opt-optimizations?view=msvc-170
+      # * Enable treating all warnings as errors, see
+      #   https://docs.microsoft.com/en-us/cpp/build/reference/wx-treat-linker-warnings-as-errors?view=msvc-170
+      # * x64 ONLY - DISABLE creation of table of safe exception handlers, see
+      #   https://docs.microsoft.com/en-us/cpp/build/reference/safeseh-image-has-safe-exception-handlers?view=msvc-170
+      # * Debug ONLY - DISABLE incremental linking, see
+      #   https://docs.microsoft.com/en-us/cpp/build/reference/incremental-link-incrementally?view=msvc-170
+      # * RelWithDebInfo ONLY - Disable incremental linking, but enable COMDAT
+      #   folding, see
+      #   https://docs.microsoft.com/en-us/cpp/build/reference/opt-optimizations?view=msvc-170
```
Comment (RytoEX) [inline cmake/external/ObsPluginHelpers.cmake:613]: Same here. Could probably drop the word "see" and just end the comment, unless there is some danger of cmake-format wrapping poorly.

File: cmake/external/ObsPluginHelpers.cmake [hunk 7]
```diff
@@ -276,37 +637,40 @@ else()
...
-        "$<$<NOT:$<EQUAL:${CMAKE_SIZEOF_VOID_P},8>>:LINKER\:/SAFESEH\:NO>"
-        "$<$<CONFIG:DEBUG>:LINKER\:/INCREMENTAL:NO>"
-        "$<$<CONFIG:RELWITHDEBINFO>:LINKER\:/INCREMENTAL:NO>")
-    endif()
-
-    setup_target_resources("${target}" "obs-plugins/${target}")
+    # Add resources from data directory
+    setup_target_resources(${target} obs-plugins/${target})
 
+    # Set up plugin for testing in available OBS build on windows
```
Comment (RytoEX) [inline cmake/external/ObsPluginHelpers.cmake:668]: windows -> Windows

**Review by RytoEX — CHANGES_REQUESTED** [review]
Some minor comment nitpicks and questions.

File: cmake/Modules/ObsDefaults_macOS.cmake [hunk 8]
```diff
@@ -14,34 +14,36 @@ if(POLICY CMP0025)
...
 # Build options
-if(NOT DEFINED CMAKE_OSX_ARCHITECTURES OR CMAKE_OSX_ARCHITECTURES STREQUAL "")
+if(NOT CMAKE_OSX_ARCHITECTURES)
   set(CMAKE_OSX_ARCHITECTURES
       "${CMAKE_HOST_SYSTEM_PROCESSOR}"
       CACHE STRING
             "OBS build architecture for macOS - x86_64 required at least" FORCE)
 endif()
-set_property(CACHE CMAKE_OSX_ARCHITECTURES PROPERTY STRINGS x86_64 arm64
+set_property(CACHE CMAKE_OSX_ARCHITECTURES PROPERTY STRINGS arm64 x86_64
```
[the remark this answers is not in the document]
Comment (PatTheMav) [reply cmake/Modules/ObsDefaults_macOS.cmake:23]: It's on purpose to mark the change of architectures to arm64 as macOS' default arch and x86_64 being the "odd one".

File: cmake/external/ObsPluginHelpers.cmake [hunk 9]
```diff
@@ -2,23 +2,23 @@ if(POLICY CMP0087)
...
   set(OS_${_SYSTEM_NAME_U} ON)
-elseif("${CMAKE_SYSTEM_NAME}" STREQUAL "Windows")
+elseif(${CMAKE_SYSTEM_NAME} STREQUAL "Windows")
   set(OS_WINDOWS ON)
   set(OS_POSIX OFF)
 endif()
 
 # Old-Style plugin detected, find "modern" libobs variant instead and set global
-# include directories to fix "bad" plugin behavior
+# include directories to fix "bad" plugin behaviour
```
[the remark this answers is not in the document]
Comment (PatTheMav) [reply cmake/external/ObsPluginHelpers.cmake]: Fixed with the other changes requested below.

File: cmake/Modules/ObsDefaults_macOS.cmake (hunk 8 above)
[the remark this answers is not in the document]
Comment (RytoEX) [reply cmake/Modules/ObsDefaults_macOS.cmake:23]: Is there any significance to the order in the string below this line?
```
"x86_64;arm64"
```

[the remark this answers is not in the document]
Comment (PatTheMav) [reply cmake/Modules/ObsDefaults_macOS.cmake:23]: I don't think it has, it's a bit of a hack, providing a "list" to CMake, but that's the way to enable universal builds (iirc CMake will translate this into a set of  `-arch` parameters being passed to clang).

[the remark this answers is not in the document]
Comment (RytoEX) [reply cmake/Modules/ObsDefaults_macOS.cmake:23]: If it's inconsequential, then it's fine.

**Review by RytoEX — APPROVED** [review]
Looks good to me.

**Pull request merged**